helix-term/src/ui/editor.rs

```diff
@@ -1461,3 +1498,8 @@ fn canonicalize_key(key: &mut KeyEvent) {
         key.modifiers.remove(KeyModifiers::SHIFT)
     }
 }
+
+pub enum DocSyntaxHighlight<'d> {
+    TreeSitter(CharacterHighlightIter<'d>),
+    Default(iter::Once<HighlightEvent>),
+}
```

Comment:
This is so that we can use static dispatch for everything, right? As opposed to a Box of a dyn trait implementation?

This enum makes the uses of `doc_syntax_highlights` awkward (looking at the usages in picker.rs and here in editor.rs) and I think what we lose in elegance of the code might not be worth the performance improvement.

Reply:
I agree that this was a bit ugly, this is one of the rare cases where the limitations of the rust type system are kind of showing (generic assoicated types would help here or generic closures in general).

I managed to implement an alternative solution in 2c6f681 that turned out quite good.
The idea is that we never return the iterator and instead call the rendering code in the same function.
An Overlay (like diagnostics, selections) are passed as parameter that implements the `HighlightOverlay`so that they are generic over the input `Iterator`. `HighlightOverlay` are composable using tuples. For example the diagnostics Layers are implemented by creating a tuple of 5 `DiagnosticOverlay`s `(hint_overlay, info_overlay, unkown_overlay, warning_overlay, error_overlay)`.